### PR TITLE
Add CRT Signers to aws-sigv4

### PR DESCRIPTION
**Pull request by alextwoods**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/s3_signer.rb [hunk 1]
```diff
@@ -175,7 +175,7 @@ def custom_endpoint?(resp)
           end
 
           def wrong_sigv4_region?(resp)
-            resp.context.http_response.status_code == 400 &&
+            [400, 301].include?(resp.context.http_response.status_code) &&
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/s3_signer.rb]: Is this intended for this change? (I typically agree we should be doing this... but maybe not as part of this)

File: gems/aws-sdk-core/spec/aws/plugins/signature_v4_spec.rb [hunk 2]
```diff
@@ -234,17 +234,16 @@ module Plugins
         }
 
         it "unsigns payload for operations has 'v4-unsigned-payload' for 'authtype'" do
+          allow(Time).to receive(:now).and_return(now)
           resp = client.streaming_foo(foo_name: 'foo')
           req = resp.context.http_request
           expect(req.headers['x-amz-content-sha256']).to eq('UNSIGNED-PAYLOAD')
-          expect(req.headers['authorization']).to eq('AWS4-HMAC-SHA256 Credential=stubbed-akid/20120101/region/svc-name/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=6d86306ea8dcf03db133cd35bcf626b0a440daefadc8ddde3304517126edb1bb')
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/signature_v4_spec.rb:240]: These tests are assuming the payload is signed, so don't we want to check for signatures?

File: gems/aws-sdk-s3/spec/object/presigned_url_spec.rb [hunk 3]
```diff
@@ -159,14 +159,14 @@ module S3
           allow(Time).to receive(:now).and_return(now)
           url, headers = obj.presigned_request(
             :get, expires_in: 86_400, request_payer: 'peccy')
-          expect(url).to eq(
+          expect(CGI.parse(url)).to eq(CGI.parse(
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/object/presigned_url_spec.rb:162]: Why do we need CGI parsing now?

File: gems/aws-sdk-s3/spec/presigner_spec.rb [hunk 4]
```diff
@@ -75,19 +75,22 @@ module S3
             bucket: 'examplebucket',
             key: 'test.txt',
             expires_in: 86_400,
-            time: Time.utc(1969, 4, 20)
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/presigner_spec.rb:78]: !

File: gems/aws-sigv4/lib/aws-sigv4/signer.rb [hunk 5]
```diff
@@ -693,8 +712,135 @@ def credentials_set?(credentials)
           !credentials.secret_access_key.empty?
       end
 
+      ### CRT Code
+
+      # the credentials used by CRT must be a
+      # CRT StaticCredentialsProvider object
+      def fetch_credentials_crt
```
Comment (mullermp) [inline gems/aws-sigv4/lib/aws-sigv4/signer.rb]: I think crt_ should be a prefix and not a suffix (at least when inspecting methods it's more obvious)

File: gems/aws-sdk-s3/spec/object/presigned_url_spec.rb (hunk 3 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/spec/object/presigned_url_spec.rb:162]: The order of the query parameters may differ.

File: gems/aws-sdk-core/spec/aws/plugins/signature_v4_spec.rb (hunk 2 above)
Comment (alextwoods) [reply gems/aws-sdk-core/spec/aws/plugins/signature_v4_spec.rb:240]: The signing is tested extensively in the `aws-sigv4` gem and there are issues with mocking here that are difficult to fix and unrelated to the functionality of unsigned-payload so I dont feel like these tests need to re-test that.

File: gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/s3_signer.rb (hunk 1 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/s3_signer.rb]: Good catch - removed from this PR.

File: gems/aws-sdk-s3/spec/presigner_spec.rb (hunk 4 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/spec/presigner_spec.rb:78]: That actually does not work in our CRT FFI bindings (dates pre-1970 will not be represented correctly).

File: gems/aws-sigv4/lib/aws-sigv4/signer.rb (hunk 5 above)
Comment (alextwoods) [reply gems/aws-sigv4/lib/aws-sigv4/signer.rb]: Sure - I don't have a preference as these are private methods.

Comment (mullermp) [reply gems/aws-sigv4/lib/aws-sigv4/signer.rb]: (even for public ones too)

**Pull request merged**